### PR TITLE
Base structure for v3 API

### PR DESCRIPTION
**Pull request by harshshah-tenable**
- This PR contains the proposed base structure for code for v3 resources
- Current structure is something like
```
io/
   __init__.py
	v3/
            __init__.py
            base/
                endpoints/
	                uw.py
                iterators/
                schema/
	                analysis.py
	                exports.py
            vm/
                api.py
                agents.py
                scanners.py
            was/
                api.py
                scanners.py
```

- Previous usage
```tio.scanners.details(50392326)```
- Updated usage
```tio.v3.vm.scanners.details(50392326)```

### CODE REVIEW

File: tenable/io/v3/__init__.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+from tenable.base.endpoint import APIEndpoint
+from .vm.api import VulnerabilityManagement
+
+class version3API(APIEndpoint):  # noqa: PLR0904
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/__init__.py:4]: Let's follow similar case convention version3API => Version3API

File: tenable/io/v3/vm/scanners.py [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+        # isn't the prettiest way to handle this, however it will consistently
+        # return the results that we are looking for.
+        def get_scanners(tmpl):
+            for item in tmpl["settings"]["basic"]["inputs"]:
+                if item["id"] == "scanner_id":
+                    return item["options"]
+            return []
+
+        vm_tmpl = self._api.policies.templates().get("advanced", None)
+        was_tmpl = self._api.policies.templates().get("was_scan", None)
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:72]: v1 WAS API.  should be deprecated.

File: tenable/io/v3/vm/scanners.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+        scanners = get_scanners(self._api.editor.template_details("scan", vm_tmpl))
+        if was_tmpl is not None:
+            scanners.extend(
+                get_scanners(self._api.editor.template_details("scan", was_tmpl))
+            )
+        return scanners
+
+    def control_scan(self, scanner_id, scan_uuid, action):
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:80]: no typing

File: tenable/io/v3/vm/scanners.py [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+            "scanners/{}/scans/{}/control".format(
+                self._check("scanner_id", scanner_id, int),
+                self._check("scan_uuid", scan_uuid, str),
+            ),
+            json={
+                "action": self._check(
+                    "action", action, str, choices=["stop", "pause", "resume"]
+                )
+            },
+        )
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:114]: _check usage should be deprecated

File: tenable/io/v3/vm/scanners.py [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+
+    def delete(self, id):
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:116]: no typing

File: tenable/io/v3/vm/scanners.py [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+    .. automethod:: get_scans
+    .. automethod:: linking_key
+    .. automethod:: search
+    .. automethod:: toggle_link_state
+"""
+# from tenable.io.v3.base import TIOEndpoint
+from tenable.io.v3.base.endpoints.uw import UWBaseEndpoint
+
+
+class ScannersAPI(UWBaseEndpoint):
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:28]: attributes `_path` and `_conv_json` should be set

File: tenable/io/v3/vm/scanners.py [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+            :obj:`None`:
+                The action was sent to the scan successfully.
+
+        Examples:
+            Stop a scan running on the scanner:
+
+            >>> tio.v3.vm.scanners.control_scan(1, '00000000-0000-0000-0000-000000000000', 'stop')
+        """
+        self._api.post(
+            "scanners/{}/scans/{}/control".format(
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:105]: should use `self._post(f'{scanner_id}/scans/{scan_uuid}/control',` instead

File: tenable/io/v3/vm/scanners.py [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+                The unique identifier for the scanner to delete.
+
+        Returns:
+            :obj:`None`:
+                The scanner was successfully deleted.
+
+        Examples:
+            >>> tio.v3.vm.scanners.delete(1)
+        """
+        self._api.delete("scanners/{}".format(self._check("id", id, int)))
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:133]: should use `self._delete(id)` instead

File: tenable/io/v3/vm/scanners.py [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+
+    def details(self, id):
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:135]: no typing

File: tenable/io/v3/vm/scanners.py [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+
+        Returns:
+            :obj:`dict`:
+                The scanner resource record.
+
+        Examples:
+            >>> scanner = tio.v3.vm.scanners.details(1)
+            >>> pprint(scanner)
+        """
+        return self._api.get("scanners/{}".format(self._check("id", id, int))).json()
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:153]: should use `self._get(id)` instead.

File: tenable/io/v3/vm/scanners.py [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+
+    def edit(self, id, **kwargs):
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:155]: typing

File: tenable/io/v3/vm/scanners.py [hunk 12]
```diff
@@ -0,0 +1,364 @@
...
+            payload["finish_update"] = 1
+        if "registration_code" in kwargs and self._check(
+            "registration_code", kwargs["registration_code"], str
+        ):
+            payload["registration_code"] = kwargs["registration_code"]
+        if "aws_update_interval" in kwargs and self._check(
+            "aws_update_interval", kwargs["aws_update_interval"], int
+        ):
+            payload["aws_update_interval"] = kwargs["aws_update_interval"]
+
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:207]: should be refactored into a marshmallow schema.

File: tenable/io/v3/vm/scanners.py [hunk 13]
```diff
@@ -0,0 +1,364 @@
...
+        self._api.put("settings/{}".format(self._check("id", id, int)), json=payload)
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:208]: _check usage should be deprecated

File: tenable/io/v3/vm/scanners.py [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+
+    def get_aws_targets(self, id):
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:210]: no typing

File: tenable/io/v3/vm/scanners.py [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+            :obj:`list`:
+                List of aws target resource records.
+
+        Examples:
+            >>> for target in tio.v3.vm.scanners.get_aws_targets(1):
+            ...      pprint(target)
+        """
+        return self._api.get(
+            "scanners/{}/aws-targets".format(self._check("id", id, int))
+        ).json()["targets"]
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:229]: should be `self._get(f'{id}/aws-targets')['targets']` instead.

File: tenable/io/v3/vm/scanners.py [hunk 16]
```diff
@@ -0,0 +1,364 @@
...
+
+    def get_scanner_key(self, id):
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:231]: no typing

File: tenable/io/v3/vm/scanners.py [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+                The scanner key
+
+        Examples:
+            >>> print(tio.v3.vm.scanners.get_scanner_key(1))
+        """
+        return str(
+            self._api.get("scanners/{}/key".format(self._check("id", id, int))).json()[
+                "key"
+            ]
+        )
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:251]: should use `str(self._get(f'{id}/key')['key'])`

File: tenable/io/v3/vm/scanners.py [hunk 18]
```diff
@@ -0,0 +1,364 @@
...
+
+    def get_scans(self, id):
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:253]: no typing

File: tenable/io/v3/vm/scanners.py [hunk 19]
```diff
@@ -0,0 +1,364 @@
...
+            :obj:`list`:
+                List of scan resource records associated to the scanner.
+
+        Examples:
+            >>> for scan in tio.v3.vm.scanners.get_scans(1):
+            ...     pprint(scan)
+        """
+        return self._api.get(
+            "scanners/{}/scans".format(self._check("id", id, int))
+        ).json()["scans"]
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:272]: should use `self._get(f'{id}/scans')['scans']`

File: tenable/io/v3/vm/scanners.py [hunk 20]
```diff
@@ -0,0 +1,364 @@
...
+
+        Returns:
+            :obj:`list`:
+                List of scanner resource records.
+
+        Examples:
+            >>> for scanner in tio.scanners.list():
+            ...     pprint(scanner)
+        '''
+        return self._api.get('scanners').json()['scanners']
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:303]: should use `self._get()['scanners']`

File: tenable/io/v3/vm/scanners.py [hunk 21]
```diff
@@ -0,0 +1,364 @@
...
+
+    def toggle_link_state(self, id, linked):
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:305]: no typing

File: tenable/io/v3/vm/scanners.py [hunk 22]
```diff
@@ -0,0 +1,364 @@
...
+
+        Examples:
+            to deactivate a linked scanner:
+
+            >>> tio.v3.vm.scanners.toggle_link_state(1, False)
+        """
+        self._api.put(
+            "scanners/{}/link".format(self._check("id", id, int)),
+            json={"link": int(self._check("linked", linked, bool))},
+        )
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:329]: should use `self._put(f'{id}/link', json={'link': int(linked)})`

File: tenable/io/v3/vm/scanners.py [hunk 23]
```diff
@@ -0,0 +1,364 @@
...
+
+    def get_permissions(self, id):
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:331]: no typing

File: tenable/io/v3/vm/scanners.py [hunk 24]
```diff
@@ -0,0 +1,364 @@
...
+            id (int): The unique identifier for the scanner.
+
+        Returns:
+            :obj:`dict`:
+                The permissions resource for the scanner
+
+        Examples:
+            >>> tio.v3.vm.scanners.get_permissions(1)
+        """
+        return self._api.permissions.list("scanner", self._check("id", id, int))
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:345]: I'd assume this should be using `self._api.v3.vm.permissions.list`


_check usage should be deprecated

File: tenable/io/v3/vm/scanners.py [hunk 25]
```diff
@@ -0,0 +1,364 @@
...
+
+    def edit_permissions(self, id, *acls):
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:347]: no typing

File: tenable/io/v3/vm/scanners.py [hunk 26]
```diff
@@ -0,0 +1,364 @@
...
+        Returns:
+            :obj:`None`:
+                The permissions have been updated successfully.
+
+        Examples:
+            >>> tio.v3.vm.scanners.edit_permissions(1,
+            ...     {'type': 'default, 'permissions': 16},
+            ...     {'type': 'user', 'id': 2, 'permissions': 16})
+        """
+        self._api.permissions.change("scanner", self._check("id", id, int), *acls)
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:364]: I'd assume that this should be using `self._api.v3.vm.permissions.change`

_check usage should be deprecated

File: tenable/io/v3/vm/scanners.py [hunk 27]
```diff
@@ -0,0 +1,364 @@
...
+    .. automethod:: control_scan
+    .. automethod:: delete
+    .. automethod:: details
+    .. automethod:: edit
+    .. automethod:: get_aws_targets
+    .. automethod:: get_scanner_key
+    .. automethod:: get_scans
+    .. automethod:: linking_key
+    .. automethod:: search
+    .. automethod:: toggle_link_state
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:22]: should be replaced with:

```python
.. rst-class:: hide-signature
.. autoclass:: ScannersAPI
    :members:
```

**Review by SteveMcGrath — CHANGES_REQUESTED** [review]

**Comment by SteveMcGrath** [comment]
When ready to merge, the commits should be squashed and likely rebased.  The duplication between here and #2 could create a marge conflict.

**Comment by harshshah-tenable** [comment]
Hi @SteveMcGrath , I have updated the PR as per your suggestions and comments.

And also Started working on test for Scanners API as per Exports

Regarding too many commits Yes this PR was raised as a draft to confirm the base v3 structure .
Once all the changes are done will make sure the commits are squashed and changes are merged with a single commit in the destination branch.

image

File: tenable/io/v3/vm/scanners.py [hunk 28]
```diff
@@ -314,7 +297,7 @@ def toggle_link_state(self, id: int, linked: bool) -> None:
 
             >>> tio.v3.vm.scanners.toggle_link_state(1, False)
         """
-        self._api.put(
+        self._put(
             "{id}/link",
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:301]: should be f-string, not normal string.

File: tenable/io/v3/vm/scanners.py [hunk 29]
```diff
@@ -291,7 +274,7 @@ def list(self) -> List:
             >>> for scanner in tio.scanners.list():
             ...     pprint(scanner)
         '''
-        return self._api.get('scanners').json()['scanners']
+        return self._get('scanners').json()['scanners']
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py]: self._get()['scanners']

File: tenable/io/v3/vm/scanners.py [hunk 30]
```diff
@@ -258,7 +241,7 @@ def get_scans(self, id: int) -> List:
             >>> for scan in tio.v3.vm.scanners.get_scans(1):
             ...     pprint(scan)
         """
-        return self._api.get(
+        return self._get(
             f"{id}/scans"
         ).json()["scans"]
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py]: no .json() when calling with _conv_json

File: tenable/io/v3/vm/scanners.py [hunk 31]
```diff
@@ -236,7 +219,7 @@ def get_scanner_key(self, id: int) -> str:
             >>> print(tio.v3.vm.scanners.get_scanner_key(1))
         """
         return str(
-            self._api.get(f"{id}/key").json()[
+            self._get(f"{id}/key").json()[
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py]: no .json() when calling with _conv_json

File: tenable/io/v3/vm/scanners.py [hunk 32]
```diff
@@ -217,7 +200,7 @@ def get_aws_targets(self, id: int) -> List:
             >>> for target in tio.v3.vm.scanners.get_aws_targets(1):
             ...      pprint(target)
         """
-        return self._api.get(f"{id}/aws-targets").json()["targets"]
+        return self._get(f"{id}/aws-targets").json()["targets"]
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:209]: no .json() when calling with _conv_json

File: tenable/io/v3/vm/scanners.py [hunk 33]
```diff
@@ -143,7 +144,7 @@ def details(self, id: int) -> Dict:
             >>> scanner = tio.v3.vm.scanners.details(1)
             >>> pprint(scanner)
         """
-        return self._api.get(f"{id}").json()
+        return self._get(f"{id}")
```
Comment (SteveMcGrath) [inline tenable/io/v3/vm/scanners.py:153]: no need for an f-string here, just passing id is sufficient.

**Review by SteveMcGrath — CHANGES_REQUESTED** [review]

File: tenable/io/v3/base/schema/export/filters.py [hunk 34]
```diff
@@ -0,0 +1,106 @@
...
+    and_ = fields.List(fields.Nested("FilterSchema"), data_key="and")
+    or_ = fields.List(fields.Nested("FilterSchema"), data_key="or")
+
+    @pre_load(pass_many=False)
+    def validate_and_transform(self, data, **kwargs):  # noqa: PLW0613
+        """
+        Handles schema validation and data transform based on the data
+        presented.
+        """
+        if (  # noqa: PLR1705
```
Comment (tomeara) [inline tenable/io/v3/base/schema/export/filters.py:24]: Might read better if you pull these conditions out into their own methods

File: tenable/io/v3/base/schema/export/filters.py [hunk 35]
```diff
@@ -0,0 +1,106 @@
...
+             'and': [
+                 {'value': '3', 'operator': 'oper', 'property': 3}
+                 ]
+             }
+        """
+        resp = {}
+        errors = {}
+        if isinstance(data, tuple):
+            oper = None
+            for elem in data:
```
Comment (tomeara) [inline tenable/io/v3/base/schema/export/filters.py:68]: I'd pull this out into its own method. `filter` would be better used here as well.

File: tenable/io/v3/vm/schema.py [hunk 36]
```diff
@@ -0,0 +1,33 @@
...
+    force_ui_update = fields.Bool()
+    finish_update = fields.Bool()
+    registration_code = fields.Str()
+    aws_update_interval = fields.Int()
+
+    @post_dump
+    def post_serialization(self, data, **kwargs):  # noqa PLR0201 PLW0613
+        data_dict = copy.deepcopy(data)
+
+        for key, value in data.items():
```
Comment (tomeara) [inline tenable/io/v3/vm/schema.py:26]: Why not use `map` or another function to iterate? I think that might be cleaner here than a `for` loop with nested logic, and an extra dict to store.

**Review by tomeara — COMMENTED** [review]
I think it mostly looks good, but there are some complex conditions that are being used that are a bit hard to follow. I would extract these out into separate methods and abstract and name them well to express what the intent is.

Also, I think some of the iteration could be better focused by using other more precise functions, rather than `for` loops with nested logic.

File: tenable/io/v3/__init__.py (hunk 1 above)
Comment (harshshah-tenable) [reply tenable/io/v3/__init__.py]: Updated.

File: tenable/io/v3/vm/scanners.py (hunk 2 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: Updated the method with the following for now.

`return NotImplementedError(
            "This method will be updated once Policies and Editor API's are implemented in v3"
        )`

File: tenable/io/v3/vm/scanners.py (hunk 3 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: Updated.

File: tenable/io/v3/vm/scanners.py (hunk 4 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: Added typing and removed _check

File: tenable/io/v3/vm/scanners.py (hunk 5 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: updated

File: tenable/io/v3/vm/scanners.py (hunk 6 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py:25]: Added

File: tenable/io/v3/vm/scanners.py (hunk 23 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: Updated

File: tenable/io/v3/vm/scanners.py (hunk 24 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: Updated the method with the following
`return NotImplementedError( "This method will be updated once Permissions API's are implemented in v3" )`

Yes it will be `self._api.v3.vm.permissions.list` once the code of Permissions API for pyTenable is implemented in v3

File: tenable/io/v3/vm/scanners.py (hunk 25 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: Updated

File: tenable/io/v3/vm/scanners.py (hunk 26 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: Updated the method with the following
`return NotImplementedError( "This method will be updated once Permissions API's are implemented in v3" )`

Yes it will be` self._api.v3.vm.permissions.list` once the code of Permissions API for pyTenable is implemented in v3

removed _check and added typing

File: tenable/io/v3/vm/scanners.py (hunk 27 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: Done.

File: tenable/io/v3/vm/scanners.py (hunk 28 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: Updated

File: tenable/io/v3/vm/scanners.py (hunk 29 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: Updated

File: tenable/io/v3/vm/scanners.py (hunk 30 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: Gotcha!

File: tenable/io/v3/vm/scanners.py (hunk 31 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: Updated

File: tenable/io/v3/vm/scanners.py (hunk 32 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: Updated

File: tenable/io/v3/vm/scanners.py (hunk 33 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: Actually the API expects a string when I tried with https://qa-develop.cloud.aws.tenablesecurity.com hence kept the string
```  File "test.py", line 65, in get_scanners
    print(tio.v3.vm.scanners.details(50392291))
  File "d:\users\hashah.ctr\desktop\tenable\pytenable\tenable\io\v3\vm\scanners.py", line 156, in details
    return self._get(id)
  File "D:\Users\hashah.ctr\Desktop\tenable\venv\lib\site-packages\restfly-1.4.4-py3.7.egg\restfly\endpoint.py", line 129, in _get
  File "D:\Users\hashah.ctr\Desktop\tenable\venv\lib\site-packages\restfly-1.4.4-py3.7.egg\restfly\endpoint.py", line 80, in _req
TypeError: sequence item 1: expected str instance, int found
```

File: tenable/io/v3/base/schema/export/filters.py (hunk 34 above)
Comment (tomeara) [reply tenable/io/v3/base/schema/export/filters.py:24]: I personally like to encapsulate complex logic like that in a separate method that can provide the evaluation, but I'm ok with adding comments to describe and refactoring later if we want

File: tenable/io/v3/vm/scanners.py [hunk 37]
```diff
@@ -0,0 +1,344 @@
...
+                The operation was requested successfully.
+
+        Examples:
+            Force a plugin update on a scanner:
+
+            >>> tio.v3.vm.scanners.edit(1, force_plugin_update=True)
+        """
+        payload = dict()
+        schema = ScannerEditSchema()
+        kwargs["id"] = id
```
[the remark this answers is not in the document]
Comment (harshshah-tenable) [reply tenable/io/v3/vm/scanners.py]: Updated the schema

File: tenable/io/v3/base/schema/export/filters.py (hunk 35 above)
Comment (harshshah-tenable) [reply tenable/io/v3/base/schema/export/filters.py]: updated the file to split the methods.

File: tenable/io/v3/base/schema/export/filters.py (hunk 34 above)
Comment (harshshah-tenable) [reply tenable/io/v3/base/schema/export/filters.py]: Added the comments in the code for filter.py

**Comment by harshshah-tenable** [comment]
@SteveMcGrath @tomeara Due to un-clean commit history in this branch raised a new PR #6 with a single commit code changes as per PR comments from this PR.

**Review by tomeara — APPROVED** [review]
Looks good. The only thing is updating those comments

File: tenable/io/v3/vm/schema.py (hunk 36 above)
Comment (harshshah-tenable) [reply tenable/io/v3/vm/schema.py]: - Updated post_dump to use map function there.
- post_dump is used as per previous code behaviour the payload was transformation was following
```
Input: 
 {
        "force_plugin_update": True,
        "force_ui_update": False,
        "finish_update": True,
        "registration_code": "random_code",
        "aws_update_interval": 60,
 }

Payload:

   {
        "force_plugin_update": 1,
        "finish_update": 1,
        "registration_code": "random_code",
        "aws_update_interval": 60,
    }
```
- False params were not sent as part of the payload and using BoolInt class will not provide us the same behaviour hence updated post_dump method accordingly

File: tenable/io/v3/base/schema/explore/analysis.py [hunk 38]
```diff
@@ -0,0 +1,16 @@
+"""
+Base Universal Workspace Analysis Schema
```
[the remark this answers is not in the document]
Comment (harshshah-tenable) [reply tenable/io/v3/base/schema/explore/analysis.py:2]: updated this one as part of #6

**Comment by harshshah-tenable** [comment]
closing this PR as #6 contains the same code and clean commit history and is merged into develop
